x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtUtil.java

```diff
@@ -0,0 +1,884 @@
+                final int proxyPort = realmConfig.getSetting(JwtRealmSettings.HTTP_PROXY_PORT);
+
+                final RequestConfig.Builder requestConfigBuilder = RequestConfig.custom();
+                requestConfigBuilder.setConnectTimeout(tcpConnectTimeoutMillis)
+                    .setConnectionRequestTimeout(tcpConnectionReadTimeoutSec)
+                    .setSocketTimeout(tcpSocketTimeout);
+                if (Strings.hasText(proxyAddress) && Strings.hasText(proxyScheme)) {
+                    requestConfigBuilder.setProxy(new HttpHost(proxyAddress, proxyPort, proxyScheme));
+                }
+                final RegistryBuilder<SchemeIOSessionStrategy> sessionStrategyBuilder = RegistryBuilder.<SchemeIOSessionStrategy>create();
+                //// final String realmConfigPrefixSslSettings = RealmSettings.realmSslPrefix(realmIdentifier);
+                // final SSLContext sslContext = sslService.sslContext(elasticsearchSslConfig);
+                // final HostnameVerifier hostnameVerifier = SSLService.getHostnameVerifier(elasticsearchSslConfig);
+                // final SSLIOSessionStrategy sslIOSessionStrategy = new SSLIOSessionStrategy(sslContext, hostnameVerifier);
+                final SSLIOSessionStrategy sslIOSessionStrategy = sslService.sslIOSessionStrategy(elasticsearchSslConfig);
+                sessionStrategyBuilder.register("https", sslIOSessionStrategy);
+
+                final PoolingNHttpClientConnectionManager httpClientConnectionManager = new PoolingNHttpClientConnectionManager(
+                    new DefaultConnectingIOReactor(),
+                    sessionStrategyBuilder.build()
+                );
+                httpClientConnectionManager.setDefaultMaxPerRoute(httpMaxEndpointConnections);
+                httpClientConnectionManager.setMaxTotal(httpMaxConnections);
+
+                final CloseableHttpAsyncClient httpAsyncClient = HttpAsyncClients.custom()
+                    .setConnectionManager(httpClientConnectionManager)
+                    .setDefaultRequestConfig(requestConfigBuilder.build())
+                    .build();
+                httpAsyncClient.start();
+                return httpAsyncClient;
+            });
+        } catch (PrivilegedActionException e) {
+            throw new IllegalStateException("Unable to create a HttpAsyncClient instance", e);
+        }
+    }
+
+    /**
+     * Use the HTTP Client to get URL content bytes up to N max bytes.
+     * @param httpClient Configured HTTP/HTTPS client.
+     * @param uri URI to download.
+     * @param maxBytes Max bytes to read for the URI. Use Integer.MAX_VALUE to read all.
+     * @return Byte array of the URI contents up to N max bytes.
+     * @throws Exception Security exception or HTTP/HTTPS failure exception.
+     */
+    @SuppressWarnings({ "removal", "unusedThrown" })
+    public static byte[] readBytes(final CloseableHttpAsyncClient httpClient, final URI uri, final int maxBytes) throws Exception {
```

Comment:
I did look at OIDC and I didn't think it could be reused. Reload logic was buried inside the OIDC ID Token validator, and we cannot reuse that validator. For example, OIDC ID Token validator does things we don't want, such as checking the nonce and accepting a single audience (i.e. the OIDC RP client_id). I have to split this out.

This helper method is structures to easily add an async thread later. First time through, the client reads the JWKSet bytes from file or URL, and saves a JWKSet object. Later, async calls will periodically re-download JWKSet bytes from the URL. Calls to authenticate() just uses whatever version is currently in memory.

Reply:
I'm not asking if we considered to reuse OIDC code, but rather the library we use in the OIDC realm, i.e. extend AbstractJWTValidator. Happy to move forward as is if you have it figured out, but I had to ask.

Reply:
Sorry I wasn't clear. I looked at OIDC realm code, and the library calls it made. I dug into the library, and it didn't seem reusable.